redash/query_runner/dgraph.py

```diff
@@ -0,0 +1,109 @@
+import json
+
+try:
+    import pydgraph
+    enabled = True
+except ImportError:
+    enabled = False
+
+from redash.query_runner import BaseQueryRunner, register
+from redash.utils import json_dumps, json_loads
+
+
+class Dgraph(BaseQueryRunner):
+    noop_query = """
+    {
+      test() {
+      }
+    }
+    """
+
+    @classmethod
+    def configuration_schema(cls):
+        return {
+            "type": "object",
+            "properties": {
+                "user": {
+                    "type": "string"
+                },
+                "password": {
+                    "type": "string"
+                },
+                "servers": {
+                    "type": "string"
+                }
+            },
+            "order": ["servers", "user", "password"],
+            "required": ["servers"],
+            "secret": ["password"]
+        }
+
+    @classmethod
+    def type(cls):
+        return "dgraph"
+
+    @classmethod
+    def enabled(cls):
+        return enabled
+
+    @classmethod
+    def annotate_query(cls):
+        """Dgraph uses '#' as a comment delimiter, not '/* */'"""
+        return False
+
+    def run_query(self, query, user):
+
+        servers = self.configuration.get('servers')
+
+        client_stub = pydgraph.DgraphClientStub(servers)
+        client = pydgraph.DgraphClient(client_stub)
+
+        txn = client.txn(read_only=True)
+        try:
+            response_raw = txn.query(query)
+
+            data = json.loads(response_raw.json)
+
+            first_key = next(iter(data.keys()))
+            first_node = data[first_key]
+
+            # grab all the column names
+            column_names = set().union(*first_node)
+            # create a dict for column names in the format that the API wants
+            # the type as 'string' is just a hack for now
+            columns = [{'name': c, 'friendly_name': c, 'type': 'string'} for c in column_names]
```

Comment:
All the values are untyped?

Reply:
> How do queries look like in Dgraph? I mean, what type of query language does it use?

Dgraph uses a variant of GraphQL they call GraphQL+- (https://docs.dgraph.io/query-language/).  The output is JSON.

There might be a syntax highlighter for GraphQL+-.    Should that be part of this PR, or a separate one?


> All the values are untyped?

Unfortunately, there's no type metadata attached to the results, like in most SQL databases.  The only way is check the JSON itself (`typeof`).

Reply:
> All the values are untyped?


Unfortunately, there's no type metadata attached to the results, like in most SQL databases.  The only way is check the JSON itself (`typeof`).

Reply:
> All the values are untyped?

Basically.  

Unfortunately, there's no type metadata attached to the results, like in most SQL databases. The only way is check the JSON itself (`typeof`).